Add a fallback mechanism after a HIBP API failure

jdufresne opened the pull request:
Previously, in the event of a HIBP API failure, the password was assumed to be valid. Instead of allowing a potentially compromised password, fallback to checking against Django's list of common passwords. While an incomplete list, the passwords are generated from the same HIBP database.

Details on Django's list of common passwords can be seen at:

https://docs.djangoproject.com/en/2.0/topics/auth/passwords/#django.contrib.auth.password_validation.CommonPasswordValidator

ubernostrum commented:
I'm not sure about this one.

First off, it feels like there's a certain amount of "self-healing" available, since if the middleware's enabled it catches the bad password on the next login attempt. So a temporary HIBP outage doesn't just permanently leave someone in the lurch with a compromised password.

Second, if this is a desired fallback behavior it can be achieved by also enabling `CommonPasswordValidator`.

Third, I worry about API compatibility; as you've implemented it, the fallback returns a `bool` indicating whether the password is in the common list, where the HIBP check returns an `int` counting the number of times the password is compromised. While those types are compatible enough for the string formatting done in the error messages, I feel like it ends up being misleading since it will always effectively say the password has been compromised once (`'%d' % True` becomes `'1'`). I'm not sure how to resolve that mismatch, or if it's possible to do so cleanly.

jdufresne commented:
Thanks for considering, James.

I understand your POV and concerns. If you ultimately decide not to merge this PR, that is ok but I'll explain why I still think it is a good idea inline.

> First off, it feels like there's a certain amount of "self-healing" available, since if the middleware's enabled it catches the bad password on the next login attempt. So a temporary HIBP outage doesn't just permanently leave someone in the lurch with a compromised password.

I've opted not to user the middleware in my project. I agree it would obviate the need for this fallback, but I find the middleware approach an invasive UX. I'd prefer to limit my project's use to the auth validator only and therefore only try to handle compromised passwords when the user is changing their password.

> Second, if this is a desired fallback behavior it can be achieved by also enabling CommonPasswordValidator.

This was my first approach, but this too creates an awkward UX. The `CommonPasswordValidator` and `PwnedPasswordsValidator` share the same help text, "Your password can't be a commonly used password". When this text is displayed to the user -- as is done with my project -- it appears as if there is a bug where the same text is displayed twice.

Additionally, these two validators are conceptually the same, with the `PwnedPasswordsValidator` being more complete by verifying the most up to date data. It feels like a small amount of waste to run the same type of validator twice. The approach I've taken here is to use `CommonPasswordValidator` only in the not-so-common-case where the HIBP API fails.

> Third, I worry about API compatibility; as you've implemented it, the fallback returns a bool indicating whether the password is in the common list, where the HIBP check returns an int counting the number of times the password is compromised. While those types are compatible enough for the string formatting done in the error messages, I feel like it ends up being misleading since it will always effectively say the password has been compromised once ('%d' % True becomes '1'). I'm not sure how to resolve that mismatch, or if it's possible to do so cleanly.

I share this concern and am open to suggestions. Here are some proposals:

* Change `return True` to `return 1`. But as you say, this lies and reports the wrong number.

* Don't catch the requests exception in `api.py`. Instead let it propagate so it can be handled by the caller. In this case, the API could avoid the fallback mechanism, but the `PwnedPasswordsValidator` (or a separate custom Validator) could handle the exception with the fallback mechanism.

mic159 commented:
I think changing the return value from timeout vs not-pwned in the API might be the best solution, eg. `None` for timeout and `False` for not-pwned.

I am considering how to add caching to this, and that distinction would be nice so we dont cache a timeout.

jdufresne commented:
> Changing the return value from timeout vs not-pwned in the API might be an acceptable solution, eg. None for timeout and False for not-pwned.

I'm open to this. But I'm curious why not just let the exception be handled by the caller as I've proposed above? The effect is mostly the same, check for `None` or check for an exception. I find the exception to be slightly nicer as it contains more information. @ubernostrum If you agree with either of these approaches, let me know and I'll go ahead.

> Then you can write your own validator that can do the fallback.

:+1:  Yup! (Or include it as a second validator in this package if that is agreeable.)

> Alternatively, your frontend could de-duplicate error messages, or only show the first one.

That only partially handles my use case:

1. It would require some slightly awkward hacks in the UI to specials case some valdiators. I'd prefer to avoid that if possible.
2. It still have two nearly duplicate validators, one that checks HIBP live data and one that checks HIBP offline data.

jdufresne commented:
I've reworked this PR based on the supplied feedback. I think the new version is more elegant and an improvement. The changes include:

* Moved the fallback mechanism from `api.py` to `validators.py`
* Fully reuse `CommonPasswordsValidator` and simply delegate to `.validate()`
* Changed `pwned_password()` to return:
  * None for API failure
  * 0 for not pwned
  * amount for pwned
* Adjusted all tests based on the above

I'm curious what you think of this new approach. All additional feedback is welcome. I'm still open to creating a second, separate validator if you prefer not to have the fallback in `PwnedPasswordsValidator`.

ubernostrum commented:
I think one big question that needs answering here is whether we expect HIBP to be down frequently enough and long enough that building robust fallback behaviors is a worthwhile endeavor. Because if it really is that unreliable, it's worth questioning whether to use it for time-sensitive things like user signups/password resets at all. And if it just blips every once in a while, then probably the fallback/error behavior can be simplified.

(I'm bringing this up not because I dislike this particular PR, but it seems like there's a precedent building of increasingly complex behavior in this app "in case HIBP is down")

jdufresne commented:
No, I don't think we HIBP will fail frequently or for long. I think it will be an uncommon edge case.

However, if HIBP API errors are handled at all, as they are now, then I think allowing users to modify their passwords to be compromised passwords is a mistake and a potential security issue of the app. So as long HIBP API failures return silently and so compromised passwords can appear as not-compromised, I think a fallback is better than not having one.

While on the subject, I think it is totally fair and useful to expect that from time to time HIBP may mistakenly return a 500, 503, or other server error. However, I think it is a bit unusual to suggest that a 200 response could be malformed. Was this even seen IRL? If we get a 200 response, I think should be confident it follows its own documented API spec. I don't think the following tests are very realistic and could perhaps be removed:

* test_bad_text
* test_bad_response_no_colon
* test_bad_response_many_colons

What do you think?

ubernostrum commented:
Do you mind updating this PR based on the other stuff that's merged?

jdufresne commented:
I've rebased on to latest master. Thanks.
